runtime/src/bank.rs

```diff
@@ -691,77 +691,106 @@ impl Bank {
             let inflation = self.inflation.read().unwrap();
 
             (*inflation).validator(year) * self.capitalization() as f64 * period
-        };
+        } as u64;
+
+        let vote_balance_and_staked = self.stakes.read().unwrap().vote_balance_and_staked();
 
-        let validator_points = self.stakes.write().unwrap().claim_points();
-        let validator_point_value =
-            self.check_point_value(validator_rewards / validator_points as f64);
+        let validator_point_value = self.pay_validator_rewards(validator_rewards);
+
+        // this sysvar could be retired...
         self.update_sysvar_account(&sysvar::rewards::id(), |account| {
             sysvar::rewards::create_account(
                 self.inherit_sysvar_account_balance(account),
                 validator_point_value,
             )
         });
 
-        let validator_rewards = self.pay_validator_rewards(validator_point_value);
+        let validator_rewards_paid =
+            self.stakes.read().unwrap().vote_balance_and_staked() - vote_balance_and_staked;
+
+        // verify that we didn't pay any more than we expected to
+        assert!(validator_rewards >= validator_rewards_paid);
```

Comment:
@aeyakovenko this is the check you wanted, I think.  we sum the vote and stake balances before and after and verify that it's less than the value of rewards

Reply:
@sakridge @ryoqun wdyt?  Does that work as a secondary computation that would have caught a similar overflow bug?

Reply:
Yeah, this should work, if we should be extremely careful (I think this is the case), I'd make `vote_balance_and_staked` return `u128` instead of `u64`. Also, I'll slightly modify the assertion like this:

```rust
    let pre_vote_balance_and_staked = ...vote_balance_and_staked();
    let max_vote_balance_and_staked = pre_vote_balance_and_staked + validator_rewards;

    // Do the fun math!
    ... self.pay_validator_rewards(validator_rewards);

    let post_vote_balance_and_staked = ...vote_balance_and_staked();

    assert!(max_vote_balance_and_staked >= post_vote_balance_and_staked && post_vote_balance_and_staked >= pre_vote_balance_and_staked);
```

The code rearrangement reason is that:
1. Don't further compute anything on possibly-contaminated (= might be already overflowing) values from riskier code (`pay_validator_rewards()`)
1. Assert the interval, not one-sided range. (This isn't perfect still, under possible overflow flaw assumption. it's still possible attacker can manage to get the crafted value into the interval. Anyway, this is better than one-sided only one `>=` assertion).

Reply:
1. if`vote_balance_and_staked()` needs to be u128, so does `bank.capitalization()`.
2. I like the range check, but could also be accomplished (possibly more concisely) with a u64::checked_add() to capitalization

Reply:
@rwalker-com

> * if`vote_balance_and_staked()` needs to be u128, so does `bank.capitalization()`.

I think `vote_balance_and_staked() -> u128` have some merit to detect otherwise possible overflow within the range of `u64` due to wrong calucation in `pay_validator_rewards`, while leaving `bank.capitalization()` to remain as `u64`. In other words, `vote_balance_and_staked() -> u128` can be justified for its sole and temporal sanity checking purpose.

> * I like the range check, but could also be accomplished (possibly more concisely) with a u64::checked_add() to capitalization

I like `u64::checked_add()` (but it's seems cumbersome to do with `AtomicU64`...)!
On the other hand, `u128` range check is still needed to detect really extreme overflows like wrapping u64::MAX twice in `pay_validator_rewards()`?

Or is it too much? Also, I think extra cpu cost due to these u128 basic math is negligible because it's just done once for an epoch.